### PR TITLE
DAOS-2824: Checksum Fault injection under VOS.

### PR DESCRIPTION
**Pull request by rpadma2**
Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

### CODE REVIEW

File: src/vos/vos_io.c [hunk 1]
```diff
@@ -394,8 +394,21 @@ akey_fetch_recx(daos_handle_t toh, const daos_epoch_range_t *epr,
...
+			if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_FETCH_FAIL)) {
+				void *fault_ptr;
+
+				D_ALLOC_ARRAY(fault_ptr,
+					csum_nr * ent->en_csum.cs_len);
+				memset(fault_ptr, 3,
+					csum_nr * ent->en_csum.cs_len);
+				memcpy(csum_ptr, fault_ptr,
+					csum_nr * ent->en_csum.cs_len);
+				D_FREE(fault_ptr);
```
Comment (wangdi1) [inline src/vos/vos_io.c]: hmm, why allocate fault_ptr here? it seems you can do
memset(csum_ptr, random(), csum_nr * ent->en_csum.cs_len) directly?

File: src/vos/vos_io.c [hunk 2]
```diff
@@ -753,8 +766,16 @@ akey_update_recx(daos_handle_t toh, daos_epoch_t epoch, uint32_t pm_ver,
...
-		ent.ei_csum = *iod_csum;
+
+	if (daos_csum_isvalid(iod_csum)) {
+		if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_UPDATE_FAIL)) {
+			ent.ei_csum = *iod_csum;
+			/* Zero the checksum field */
+			memset(ent.ei_csum.cs_csum, 0, sizeof(uint8_t));
+		} else {
+			ent.ei_csum = *iod_csum;
+		}
```
Comment (wangdi1) [inline src/vos/vos_io.c]: it seems ent.ei_csum = *iod_csum; can be moved outside of if() else {}

File: src/vos/tests/csum_extent_tests.c [hunk 3]
```diff
@@ -942,3 +942,77 @@ void csum_invalid_input_tests(void **state)
...
+			fail_msg("Error updating extent with csum: %s\n",
+				 d_errstr(rc));
+
+		daos_fail_loc_set(DAOS_CHECKSUM_FETCH_FAIL | DAOS_FAIL_ONCE);
+		rc = fetch(&test, fetch_extents, &csum_count_per_extent,
+			   &csum_len, &csums_count_total, i);
+		if (!SUCCESS(rc))
+			fail_msg("Error fetching extent with csum: %s\n",
+				 d_errstr(rc));
+		daos_fail_loc_set(0);
```
Comment (wangdi1) [inline src/vos/tests/csum_extent_tests.c]: you can cleanup fail_loc after the loop.

File: src/vos/tests/csum_extent_tests.c [hunk 4]
```diff
@@ -942,3 +942,77 @@ void csum_invalid_input_tests(void **state)
...
+
+	/* Setup Test */
+	struct csum_test_params params;
+
+	params.total_records = 1024 * 1024 * 64;
+	params.record_bytes = 1;
+	params.csum_bytes = 8; /* CRC64? */
+	params.csum_chunk_records = 1024 * 16; /* 16K */
+	params.use_rand_csum = true;
+	struct csum_test test;
```
[the remark this answers is not in the document]
Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: Done. Will be in next patch.

File: src/vos/vos_io.c (hunk 1 above)
Comment (rpadma2) [reply src/vos/vos_io.c]: Thanks. Done.

File: src/vos/vos_io.c (hunk 2 above)
Comment (rpadma2) [reply src/vos/vos_io.c]: Done.

File: src/vos/tests/csum_extent_tests.c (hunk 3 above)
Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: Done.

File: src/vos/tests/csum_extent_tests.c [hunk 5]
```diff
@@ -942,3 +942,78 @@ void csum_invalid_input_tests(void **state)
...
 	free(csum_buf_1);
 }
+
+/* Inject fault on the update and fetch.
+* The update will initialize checksum value to zero.
+* Fetch operation will return random values in checksum field.
+* Test will compare the update and fetch output for pass/fail
+*/
+void
+csum_fault_injection_multiple_extents_tests(void **state)
```
Comment (ryon-jensen) [inline src/vos/tests/csum_extent_tests.c]: Does this test verify more logic than what's already tested in your changes to io_update_and_fetch_dkey? Honestly, I would rather have more dedicated tests like this than adding additional complexity to other tests, but io_update_and_fetch_dkey does cover single values as well. Should we add specific test cases for single value checksums? Thoughts?

File: src/vos/tests/csum_extent_tests.c [hunk 6]
```diff
@@ -942,3 +942,78 @@ void csum_invalid_input_tests(void **state)
...
+			fail_msg("Error fetching extent with csum: %s\n",
+				 d_errstr(rc));
+
+		/* Verify */
+		assert_int_equal(test.csum_bytes, csum_len);
+		assert_int_equal(get_csums_per_extent(&test, fetch_extents),
+				 csum_count_per_extent);
+		assert_int_equal(get_csum_total(&test, fetch_extents),
+				csums_count_total);
+		assert_memory_not_equal(test.update_csum_buf,
```
Comment (ryon-jensen) [inline src/vos/tests/csum_extent_tests.c]: I don't think this fully verifies the test here. It does verify the checksum is corrupted, but does it know if it happened on update or fetch? If the error inject on update doesn't actually work, the fetch error inject would hide that and vice versa. Or am I missing something?

File: src/vos/tests/csum_extent_tests.c (hunk 5 above)
Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: The io_update_fetch_dkey single value coverage should be fine for now. In future, we can add separate test case under csum_extent_tests.c file. All changes in vos_io.c should be able still handle injecting faults.

File: src/vos/tests/csum_extent_tests.c (hunk 6 above)
Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: The problem here is all tests are running on the VOS. I believe when we have the client/server model fully functional, we should be able to create tests with update fault and see whether fetch gives different value. Presently, all the tests are running on VOS (even if we inject fault on update, fetch is going to get the bad data unless we can tell what fetch is supposed get as value). Maybe when we calculate actual checksum under server, we can enhance/add the test cases. My few cents.

File: src/vos/vos_io.c [hunk 7]
```diff
@@ -753,8 +776,13 @@ akey_update_recx(daos_handle_t toh, daos_epoch_t epoch, uint32_t pm_ver,
 	ent.ei_rect.rc_ex.ex_hi = recx->rx_idx + recx->rx_nr - 1;
 	ent.ei_ver = pm_ver;
 	ent.ei_inob = rsize;
-	if (daos_csum_isvalid(iod_csum))
+
+	if (daos_csum_isvalid(iod_csum)) {
 		ent.ei_csum = *iod_csum;
+		/* Zero the checksum field for fault injection*/
+		if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_UPDATE_FAIL))
+			memset(ent.ei_csum.cs_csum, 0, sizeof(uint8_t));
```
[the remark this answers is not in the document]
Comment (rpadma2) [reply src/vos/vos_io.c]: Done.

File: src/vos/vos_io.c [hunk 8]
```diff
@@ -279,7 +279,20 @@ akey_fetch_single(daos_handle_t toh, const daos_epoch_range_t *epr,
...
 	tree_rec_bundle2iov(&rbund, &riov);
 	rbund.rb_biov	= &biov;
-	rbund.rb_csum	= &iod->iod_csums[0];
+
+	if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_FETCH_FAIL)) {
+		/* Get the iod_csum pointer and
+		* manipulate the checksum value
+		* for fault injection.
+		*/
+		rbund.rb_csum	= &iod->iod_csums[0];
```
[the remark this answers is not in the document]
Comment (rpadma2) [reply src/vos/vos_io.c]: Done. Thanks.

File: src/vos/tests/csum_extent_tests.c (hunk 6 above)
Comment (jolivier23) [reply src/vos/tests/csum_extent_tests.c]: I'm not sure I grok the reasoning but it's Friday afternoon.

**Comment by jolivier23** [comment]
Can you merge latest master?

**Comment by rpadma2** [comment]
> Can you merge latest master?

Completed the merge with master.

**Comment by rpadma2** [comment]
Updated DAOS_FAIL_ONCE with DAOS_FAIL_ALWAYS flag... Looks like once in a while the flag is not set properly if we use DAOS_FAIL_ONCE in a while loop.

**Comment by rpadma2** [comment]
Looks like the problem is fixed after changing fault injection flag from DAOS_FAIL_ONCE to DAOS_FAIL_ALWAYS. I tried to run_test.sh on my test machine and noticed there is no intermittent failures.

File: src/vos/tests/csum_extent_tests.c [hunk 9]
```diff
@@ -942,3 +942,80 @@ void csum_invalid_input_tests(void **state)
...
+
+		daos_fail_loc_set(DAOS_CHECKSUM_UPDATE_FAIL | DAOS_FAIL_ALWAYS);
+		daos_fail_num_set(1);
+		rc = update(&test, update_extents, i);
+		if (!SUCCESS(rc))
+			fail_msg("Error updating extent with csum: %s\n",
+				 d_errstr(rc));
+
+		daos_fail_loc_set(DAOS_CHECKSUM_FETCH_FAIL | DAOS_FAIL_ALWAYS);
+		daos_fail_num_set(1);
```
Comment (wangdi1) [inline src/vos/tests/csum_extent_tests.c]: what is this fail_num_set(1) for?

File: src/vos/tests/vts_io.c [hunk 10]
```diff
@@ -682,6 +687,10 @@ io_update_and_fetch_dkey(struct io_test_args *arg, daos_epoch_t update_epoch,
 	iod.iod_nr	= 1;
 
 	/* Act */
+	if (fault_injection_flag == FAULT_INJECT) {
+		daos_fail_loc_set(DAOS_CHECKSUM_UPDATE_FAIL | DAOS_FAIL_ALWAYS);
+		daos_fail_num_set(2);
```
Comment (wangdi1) [inline src/vos/tests/vts_io.c]: same here, what is this for?

File: src/vos/tests/csum_extent_tests.c (hunk 9 above)
Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: Based on my testing, it is the number of times fault injection flag is tested. 
eg: If we perform DAOS_FAIL_CHECK once, the fail_number is decremented by 1.  
In multiple extent testing, we are doing the DAOS_FAIL_CHECK once. In the single value testing, we are doing DAOS_FAIL_CHECK at twice.

File: src/vos/tests/vts_io.c (hunk 10 above)
Comment (rpadma2) [reply src/vos/tests/vts_io.c]: Same comment:

Based on my testing, it is the number of times fault injection flag is tested.
eg: If we perform DAOS_FAIL_CHECK once, the fail_number is decremented by 1.
In multiple extent testing, we are doing the DAOS_FAIL_CHECK once. In the single value testing, we are doing DAOS_FAIL_CHECK at twice.

File: src/vos/tests/csum_extent_tests.c (hunk 9 above)
Comment (wangdi1) [reply src/vos/tests/csum_extent_tests.c]: Well, FAIL_ALWAYS will always set the failure injection, until fail_loc_reset(). 
Actually, fail_num_set() is used together with DAOS_FAIL_SOME. Just to be clear, 

fail_loc_set(.....DAOS_FAIL_SOME);
daos_fail_num_set(1)

equals to
fail_loc_set(DAOS_FAIL_ONCE).

Comment (rpadma2) [reply src/vos/tests/csum_extent_tests.c]: Oh, I see. Thanks. Let me remove the daos_fail_set and use just DAOS_FAIL_ALWAYS alone and re-run the testing.

**Pull request merged**